trestle/core/remote/cache.py

```diff
@@ -0,0 +1,321 @@
+                'Cache request for invalid input URI:'
+                f'Attempt to cache from location within a trestle project {self._uri}'
+            )
+        shutil.copy(self._abs_path, self._inst_cache_path)
+
+
+class SFTPFetcher(FetcherBase):
+    """Fetcher for SFTP content."""
+
+    def __init__(self, trestle_root: pathlib.Path, uri: str, refresh: bool = False, cache_only: bool = False) -> None:
+        """Initialize SFTP fetcher. Update the expected cache path as per caching specs.
+
+        Arguments:
+        ----------
+        trestle_root: pathlib.Path
+            Path of the Trestle project path, i.e., within which .trestle is to be found.
+        uri: str
+            Reference to the remote file to cache that can be fetched using the sftp:// scheme.
+        refresh: bool
+            Whether or not the cache should be refreshed
+        cache_only: bool
+            Whether or not the operation should only target the cache copy
+        """
+        super().__init__(trestle_root, uri, refresh, cache_only)
+        # Is this a valid uri, however? Username and password are optional, of course.
+        u = parse.urlparse(self._uri)
+        if not u.hostname:
+            logger.error(f'Malformed URI, cannot parse hostname in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing hostname {self._uri}')
+        if not u.path:
+            logger.error(f'Malformed URI, cannot parse path in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing file path {self._uri}')
+
+        sftp_cached_dir = self._trestle_cache_path / u.hostname
+        # Skip any number of back- or forward slashes preceding the url path (u.path)
+        path_parent = pathlib.Path(u.path[re.search('[^/\\\\]', u.path).span()[0]:]).parent
+        sftp_cached_dir = sftp_cached_dir / path_parent
+        sftp_cached_dir.mkdir(parents=True, exist_ok=True)
+        self._inst_cache_path = sftp_cached_dir / pathlib.Path(pathlib.Path(u.path).name)
+
+    def _sync_cache(self) -> None:
+        """Fetch remote object and update the cache if appropriate and possible to do so.
+
+        Authentication relies on the user's private key being either active via ssh-agent or
+        supplied via environment variable SSH_KEY. In the latter case, it must not require a passphrase prompt.
+        """
+        u = parse.urlparse(self._uri)
+        client = paramiko.SSHClient()
+        # Must pick up host keys from the default known_hosts on this environment:
+        try:
+            client.load_system_host_keys()
+        except Exception as e:
+            logger.error('Error loading system host keys.')
+            logger.debug(e)
+            raise TrestleError(f'Cache update failure for {self._uri}')
+        # Use the supplied private key file if given, or look for keys in default path.
+        if 'SSH_KEY' in os.environ:
+            key_filename = os.environ['SSH_KEY']
```

Comment:
Are you passing a key here? or a key file name? My read on the functionality for paramiko is you should be *DIRECTLY* passing the ssh key to the pkey function or changing the doucmentation. Both cannot be right.

Reply:
Done.